lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp

```diff
@@ -0,0 +1,58 @@
+#include "lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.h"
+
+#include <llvm/Support/raw_ostream.h>
+
+#include "lib/Dialect/Secret/IR/SecretOps.h"
+#include "lib/Dialect/Secret/IR/SecretTypes.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"      // from @llvm-project
+#include "mlir/include/mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"        // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"            // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"         // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"        // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+void SecretnessAnalysis::setToEntryState(SecretnessLattice *lattice) {
+  auto operand = lattice->getPoint();
+  bool isSecret = isa<secret::SecretType>(operand.getType());
+
+  Operation *operation = nullptr;
+  // Get defining operation for operand
+  if (auto blockArg = dyn_cast<BlockArgument>(operand)) {
+    operation = blockArg.getOwner()->getParentOp();
+  } else {
+    operation = operand.getDefiningOp();
+  }
+
+  // If operand is defined by a secret.generic operation, check if operand is of
+  // secret type
+  if (auto genericOp = dyn_cast<secret::GenericOp>(*operation)) {
+    if (OpOperand *genericOperand =
+            genericOp.getOpOperandForBlockArgument(operand)) {
+      isSecret = isa<secret::SecretType>(genericOperand->get().getType());
+    }
+  }
+  propagateIfChanged(lattice, lattice->join(Secretness(isSecret)));
+}
+
+void SecretnessAnalysis::visitOperation(
+    Operation *operation, ArrayRef<const SecretnessLattice *> operands,
+    ArrayRef<SecretnessLattice *> results) {
+  for (const SecretnessLattice *operand : operands) {
```

Comment:
I think this nested loop would be better decomposed into two separate loops:

Loop 1: loop over operands and check if the lattice for any is secret, accumulating into a single Boolean.
Loop 2: loop over the results and join+propagate with the result of loop 1.

If you wanted to be slightly more generic, the accumulator operation in loop 1 can be `join`, rather than knowing about the internal structure of the lattice.